### DISCUSSION TITLE
We must catchup -> Too much latency! Resetting active sources

### DISCUSSION THREAD

**Discussion by gavin2812** [savonet/liquidsoap#3663] [User Support]
Hi all,

I'm trying to find a solution to the following:

```
2024/01/23 00:00:22 [clock.main:2] We must catchup 51.81 seconds!
2024/01/23 00:00:24 [clock.main:2] We must catchup 53.29 seconds!
2024/01/23 00:00:26 [clock.main:2] We must catchup 55.99 seconds!
2024/01/23 00:00:28 [clock.main:2] We must catchup 57.17 seconds!
2024/01/23 00:00:29 [clock.main:2] We must catchup 58.18 seconds!
2024/01/23 00:00:31 [clock.main:2] Too much latency! Resetting active sources...
2024/01/23 00:00:31 [mksafe:3] Switch to safe_blank with forgetful transition.
2024/01/23 00:00:31 [replay_metadata_7:5] Activations changed: static=[], dynamic=[].
2024/01/23 00:00:31 [source:4] Source replay_metadata_7 gets down.
2024/01/23 00:00:31 [input.http_0:5] Activations changed: static=[input.http_0], dynamic=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0].
2024/01/23 00:00:31 [safe_blank:5] Activations changed: static=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0], dynamic=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0].
2024/01/23 00:00:33 [clock.main:2] We must catchup 1.18 seconds!
2024/01/23 00:00:34 [clock.main:2] We must catchup 2.75 seconds!
2024/01/23 00:00:35 [clock.main:2] We must catchup 3.68 seconds!
2024/01/23 00:00:36 [decoder.ffmpeg:4] ffmpeg recognizes "https:domain.com/stream" as: audio: {codec: aac, 44100Hz, 2 channel(s)} and content-type: {audio=pcm(stereo),video=none,midi=none}.
2024/01/23 00:00:36 [mksafe:3] Switch to input.http_0 with transition.
2024/01/23 00:00:36 [safe_blank:5] Activations changed: static=[], dynamic=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0].
2024/01/23 00:00:36 [source:4] Source replay_metadata_8 gets up with content kind: {audio=pcm(stereo),video=none,midi=none}.
2024/01/23 00:00:36 [replay_metadata_8:5] Clock is main[].
2024/01/23 00:00:36 [replay_metadata_8:5] Content kind: {audio=pcm(stereo),video=none,midi=none}, content type: {audio=pcm(stereo),video=none,midi=none}
2024/01/23 00:00:36 [replay_metadata_8:4] Content type is {audio=pcm(stereo),video=none,midi=none}.
2024/01/23 00:00:36 [input.http_0:5] Activations changed: static=[replay_metadata_8:mksafe:mksafe:buffer.consumer_0:buffer.consumer_0, input.http_0], dynamic=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0].
2024/01/23 00:00:36 [replay_metadata_8:5] Activations changed: static=[mksafe:mksafe:buffer.consumer_0:buffer.consumer_0], dynamic=[].
2024/01/23 00:00:37 [input.http_0:5] Got metadata at position 0: calling handlers...
2024/01/23 00:00:37 [replay_metadata_8:5] Got metadata at position 0: calling handlers...
2024/01/23 00:00:37 [mksafe:5] Got metadata at position 0: calling handlers...
2024/01/23 00:00:37 [mksafe:5] Got metadata at position 0: calling handlers...
2024/01/23 00:00:37 [clock.main:2] We must catchup 5.44 seconds!
2024/01/23 00:00:40 [clock.main:2] We must catchup 7.75 seconds!
2024/01/23 00:00:41 [clock.main:2] We must catchup 8.90 seconds!
2024/01/23 00:00:42 [clock.main:2] We must catchup 9.71 seconds!
```

All the liquidsoap script does it take the incoming stream and send it out to an Icecast server as follows:

```
output.icecast(%mp3(bitrate=128,samplerate=44100),
  host="host.com", 
  description="Stream", 
  genre="Pop",
  name="Stream",
  url=" Stream",
  port=8000, 
  password="hackme",
  mount="stream",
  fallible=true,
  mksafe(input))
    )
```

Each time this happens the output disconnects which is problematic. It will reconnect again, but the output has already dropped and this causes a problem further down the chain.

I understand that the input stream is problematic and 99.999% causing the issue, unfortunately in this instance I cannot influence the input stream, but what I'd like to do is find a solution whereby the Icecast server does not disconnect when this happens, maybe play a silence while it reconnects, or similar.

My questions are...

1) Can I use something like a fallback and play a silence, would this leave the output connected?
2) Could I maybe configure a second audio stream (same URL) and just use two identical input streams in a fallback?
3) I have read about clocks, but unfortunately I do not know how to implement this OR if this would be a suitable solution.
4) Is there another scenario I could look at for this?

Ultimately the goal is that the output Icecast stays online and does not disconnect whatever the situation with the input stream.

I'm not that proficient with liquidsoap so please be gentle!

Many thanks in advance.

**Comment by vitoyucepi** [comment]
Hi @gavin2812,
> Can I use something like a fallback and play a silence, would this leave the output connected?

`mksafe(input)` is a fallback to silence.

> Is there another scenario I could look at for this?

1. Maybe a VPN like wireguard?
2. Large buffer, or `buffer.adaptive`.

> input stream is problematic and 99.999% causing the issue, unfortunately in this instance I cannot influence the input stream

Is this a problem in general or just when pulling from liquidsoap?

**Reply by gavin2812** [reply]
@vitoyucepi thanks for your reply, sorry for delay.

Where can I find info on buffer.adaptive I tried looking/search but cannot find it documented?

How would I implement it in this input scenario?

```
#Input
url = "http://mydomain.com/stream"
live = input.http(url)
```
Thanks!

**Reply by vitoyucepi** [reply]
> input stream is problematic and 99.999% causing the issue, unfortunately in this instance I cannot influence the input stream

Is this a problem in general or just when pulling from liquidsoap?

> Where can I find info on buffer.adaptive I tried looking/search but cannot find it documented?

- [`buffer`](https://www.liquidsoap.info/doc-2.2.4/reference.html#buffer)
- [`buffer.adaptive`](https://www.liquidsoap.info/doc-2.2.4/reference.html#buffer.adaptative)

Also, take a look at pages 202 and 253 of [the book](https://www.liquidsoap.info/book/book.pdf). Or just search for the term `buffer'.

> How would I implement it in this input scenario?
> ```ruby
> url = "http://mydomain.com/stream"
> live = input.http(url)
> ```

```ruby
live = buffer.adaptive(live, buffer=30., averaging=60., max=90.)
# live = buffer(live, buffer=10., max=60.)
```

Maybe I don't know something and what we're looking for is #3744.
```ruby
live = input.http(url, self_sync=false)
```

**Reply by gavin2812** [reply]
@vitoyucepi Thanks!
